docs: improve partitioning section (#1079)
neo4j-contrib/neo4j-spark-connector: docs/modules/ROOT/pages/performance/spark.adoc

== Partitioning

> When the connector fetches data from Neo4j during read operations we offer the possibility to partition the query in smaller partitions.

Reviewer: ```suggestion
When the connector fetches data from Neo4j during read operations  you can partition the query.
```

But a good idea to say why you might want to partition at this point. Is it just to run parallel operations?

> Each individual partition will then run as separate Spark tasks, which in turn run in parallel.

Reviewer: ```suggestion
The partitions run in parallel as separate Spark tasks.
```

This means that if the total count of the nodes with label `Person` into Neo4j is 100 we are creating 5
partitions and each one manages 20 records (we use `SKIP / LIMIT` queries).

> Partitioning the dataset makes sense only if you're dealing with a big dataset (rule of thumb approximately >= 10M of records).

Reviewer: ```suggestion
Partitioning the dataset makes sense when  you are working with a large dataset, for example where you have more than 10 million records.
```

[#parallelize]
=== How to parallelize query execution

Partitions option is supported by three of our modes.
GDS is not supported by this option.
You can therefore partition these modes:

1. Node extraction.
2. Relationship extraction.
3. Query extraction.

> A dynamic count on what you are trying to fetch is being provided and a query with `SKIP / LIMIT` approach over each partition is being built.

Reviewer: ```suggestion
A dynamic count on what you are trying to fetch is provided and a query with `SKIP / LIMIT` approach over each partition is built.
```

> A dataset of 100 nodes (`Person`) with a partition size of 5 the following queries are generated (one for partition):

Reviewer: ```suggestion
For a dataset of 100 nodes (`Person`) with a partition size of 5 the following queries are generated (one for each partition):
```

----

For node and relationship extractions, the connector will automatically leverage the link:https://neo4j.com/developer/kb/fast-counts-using-the-count-store/[Neo4j count store] in order to retrieve the total count of the nodes and/or relationships you are trying to fetch.
However, for custom query extractions, you will have to provide an additional optimized count query, or a static number if you already know the count beforehand.

[NOTE]
====
The count query is executed once and should be considered as additional overhead for when partitions is enabled.
It is just as important to optimize your count query as it is to optimize the actual query.
By providing a custom count query, you gain the flexibility to optimize it against your domain.
====

The connector will opt for one of three possible approaches to determine a global partition count.
By setting option `query.count` you influence what approach is being used:

1. Compute the global count using a second *optimized* query that leverages indexes.
  * Do this by setting `.option("query.count", "<your cypher count query>")`.
2. Provide a static integer value as count in case you know it beforehand.

>   * Do this by setting option `query.count` to an integer, e.g. `.option("query.count", 125000000)`

Reviewer 2: nit: 125000000 is a number is a lot or 0s. Maybe we can give a simpler number as an example? 100 in the previous example was fine by me. 20 could be aligned with the example above.

Author: aligning with example above. Which would be 100 because total count not partition size

3. If you do not set any value to `query.count` we will fall back and automatically generate one.
  * Generated on the form of `CALL { <your query> } RETURN count(*) AS count` for you.

A count query must return a single integer as `count`, consider this example:

[source,cypher]
----
MATCH (p:Person)-[r:BOUGHT]->(pr:Product)
WHERE pr.name = 'An Awesome Product'
RETURN count(p) AS count
----

Failing to provide such a query or count will fall back to approach (3) which has no guarantee to be an optimized query.
It is therefore strongly recommended to use indexes and custom count queries in favor of falling back to (3).

> === Deterministic ordering of partitions

Reviewer 2: Let's also add a warning about the possibility of duplicate or missed reads if source database is under heavy writes that change the total number of rows due to how partitions are calculated.

Author: I struggled to word it to be honest, but it's in the latest version now. Please have a look lmk how it can be improved

MATCH (p:Person) RETURN p SKIP 80 LIMIT 20
----

> For node and relationship extractions, the connector will automatically leverage the link:https://neo4j.com/developer/kb/fast-counts-using-the-count-store/[Neo4j count store] in order to retrieve the total count of the nodes and/or relationships you are trying to fetch.

Reviewer: ```suggestion
For node and relationship extractions, the connector automatically leverages the link:https://neo4j.com/developer/kb/fast-counts-using-the-count-store/[Neo4j count store] in order to retrieve the total count of the nodes and/or relationships you are trying to fetch.
```

> However, for custom query extractions, you will have to provide an additional optimized count query, or a static number if you already know the count beforehand.

Reviewer: ```suggestion
However, for custom query extractions, you need to provide an additional optimized count query, or a static number if you already know the count.
```

[NOTE]
====
The count query is executed once and should be considered as additional overhead for when partitions is enabled.
It is just as important to optimize your count query as it is to optimize the actual query.
By providing a custom count query, you gain the flexibility to optimize it against your domain.
====

The connector will opt for one of three possible approaches to determine a global partition count.

> By setting option `query.count` you influence what approach is being used:

Reviewer: ```suggestion
By setting option `query.count` you influence the approach that is used:
```

1. Compute the global count using a second *optimized* query that leverages indexes.
  * Do this by setting `.option("query.count", "<your cypher count query>")`.

> 2. Provide a static integer value as count in case you know it beforehand.

Reviewer: ```suggestion
2. Provide a static integer value as count if you know the value.
```

  * Do this by setting option `query.count` to an integer, e.g. `.option("query.count", 100)`

> 3. If you do not set any value to `query.count` we will fall back and automatically generate one.

Reviewer: ```suggestion
3. If you do not set a value for `query.count` a value is automatically generated.
```

  * Generated on the form of `CALL { <your query> } RETURN count(*) AS count` for you.

A count query must return a single integer as `count`, consider this example:

[source,cypher]
----
MATCH (p:Person)-[r:BOUGHT]->(pr:Product)
WHERE pr.name = 'An Awesome Product'
RETURN count(p) AS count
----

Failing to provide such a query or count will fall back to approach (3) which has no guarantee to be an optimized query.

> It is therefore strongly recommended to use indexes and custom count queries in favor of falling back to (3).

Reviewer: I found this section a little confusing at first. I think the cypher query is an example of doing option (1) above, but I'm not sure.

It might be worth turning the 'Failing to provide' content into an admonition, such as a `[TIP]`

Reviewer 3: Applied in 843e347.

=== Deterministic ordering of partitions

In order to guarantee that there is no data overlap between the partitions, you will have to provide a deterministic ordering in your query.
Simplest way to do this is to be explicit with the ordering using `ORDER BY` in your query.

> Failing to do so can lead to incomplete extractions, where some data is never included in *any* of the partitions.

Reviewer: ```suggestion
To guarantee that there is no data overlap between the partitions, you must provide a deterministic ordering in your query.
To do this,  use `ORDER BY` in your query to  provide explicit ordering.
If you do not provide an explicit order, it can lead to incomplete extractions, where some data is never included in *any* of the partitions.
```

This is a documented Cypher behavior, please see link:https://neo4j.com/docs/cypher-manual/current/clauses/skip[SKIP] in the Cypher manual for more details.

=== Reading partitions when under heavy write-load

The partition optimization feature assumes that it is safe to run each individual partition of work in their own driver transactions.

> Which is not the case if your instance is under heavy write-load.

Reviewer: ```suggestion
This is not the case if your instance is under heavy write-load.
```

When your database is under heavy write-load it sees frequent state changes.

> The count query runs in one transaction before creating the partitions, whom will all run under induvidual transactions.

Reviewer: ```suggestion
The count query runs in one transaction before creating the partitions, which then all run under individual transactions.
```

> Any data written after partitions are determined and before actual reading operation manage to start execution will not be read.

Reviewer 2: I've replace `will not be read`, `may not be read` as this behaviour may not be deterministic.

Author: Thank you, I'll send this off to team docs
